### PR TITLE
Patch containing TLS implementation for GOST 2012

### PR DESCRIPTION
**Pull request by beldmit**
This is a preliminary patch containing necessary changes providing GOST 2012 ciphersuites in TLS.

### CODE REVIEW

File: ssl/s3_srvr.c [hunk 1]
```diff
@@ -2761,8 +2761,24 @@ int ssl3_get_client_key_exchange(SSL *s)
...
+            if (pk == NULL) {
+                pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
+            }
+        } else if (alg_a & SSL_aGOST12) {
+            pk = s->cert->pkeys[SSL_PKEY_GOST12_512].privatekey;
+            if (pk == NULL) {
+                pk = s->cert->pkeys[SSL_PKEY_GOST12_256].privatekey;
+            }
+        }
+
```
Comment (mattcaswell) [inline ssl/s3_srvr.c]: As I understand it from what you sent me the spec says you can only use GOST01 keys for GOST01 ciphers. For GOST12 ciphers you can use either GOST01 or GOST12 keys - right? In addition you have set the GOST01 bit for all GOST12 ciphers. So this logic doesn't quite do that I think. Isn't it supposed to look more like this?

````
        if (alg_a & SSL_aGOST12) {
            pk = s->cert->pkeys[SSL_PKEY_GOST12_512].privatekey;
            if (pk == NULL) {
                pk = s->cert->pkeys[SSL_PKEY_GOST12_256].privatekey;
            }
            if (pk == NULL) {
                pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
            }
        } else if (alg_a & SSL_aGOST01) {
            pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
        }
````

File: ssl/s3_srvr.c [hunk 2]
```diff
@@ -2903,8 +2919,11 @@ int ssl3_get_cert_verify(SSL *s)
      * If key is GOST and n is exactly 64, it is bare signature without
      * length field
      */
-    if (n == 64 && pkey->type == NID_id_GostR3410_2001) {
-        len = 64;
+    if (n == 64 && (pkey->type == NID_id_GostR3410_2001 ||
+                    pkey->type == NID_id_GostR3410_2012_256)) {
+         len = 64;
+    } else if (n == 128 && (pkey->type == NID_id_GostR3410_2012_512)) {
+         len = 128;
```
Comment (mattcaswell) [inline ssl/s3_srvr.c]: Please check this workaround is actually needed for GOST12

File: ssl/s3_srvr.c (hunk 1 above)
Comment (beldmit) [inline ssl/s3_srvr.c]: Hello Matt!

On Mon, Sep 21, 2015 at 10:38 PM, mattcaswell <notifications@github.com>
wrote:

> In ssl/s3_srvr.c
> <https://github.com/openssl/openssl/pull/407#discussion_r40015358>:
>
> > +             * New GOST ciphersuites have SSL_aGOST01 bit too
> > +             */
> > +            pk = s->cert->pkeys[SSL_PKEY_GOST12_512].privatekey;
> > +            if (pk == NULL) {
> > +                pk = s->cert->pkeys[SSL_PKEY_GOST12_256].privatekey;
> > +            }
> > +            if (pk == NULL) {
> > +                pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
> > +            }
> > +        } else if (alg_a & SSL_aGOST12) {
> > +            pk = s->cert->pkeys[SSL_PKEY_GOST12_512].privatekey;
> > +            if (pk == NULL) {
> > +                pk = s->cert->pkeys[SSL_PKEY_GOST12_256].privatekey;
> > +            }
> > +        }
> > +
>
> As I understand it from what you sent me the spec says you can only use
> GOST01 keys for GOST01 ciphers. For GOST12 ciphers you can use either
> GOST01 or GOST12 keys - right? In addition you have set the GOST01 bit for
> all GOST12 ciphers. So this logic doesn't quite do that I think. Isn't it
> supposed to look more like this?
>
>         if (alg_a & SSL_aGOST12) {
>             pk = s->cert->pkeys[SSL_PKEY_GOST12_512].privatekey;
>             if (pk == NULL) {
>                 pk = s->cert->pkeys[SSL_PKEY_GOST12_256].privatekey;
>             }
>             if (pk == NULL) {
>                 pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
>             }
>         } else if (alg_a & SSL_aGOST01) {
>             pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
>         }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/407/files#r40015358>.
>

I asked my former colleagues to make extra checks.I will report as soon as
I get the result.

-- 
SY, Dmitry Belyavsky

File: ssl/s3_srvr.c (hunk 2 above)
Comment (beldmit) [inline ssl/s3_srvr.c]: On Mon, Sep 21, 2015 at 10:42 PM, mattcaswell <notifications@github.com>
wrote:

> In ssl/s3_srvr.c
> <https://github.com/openssl/openssl/pull/407#discussion_r40015849>:
>
> > @@ -2903,8 +2919,11 @@ int ssl3_get_cert_verify(SSL *s)
> >       * If key is GOST and n is exactly 64, it is bare signature without
> >       * length field
> >       */
> > -    if (n == 64 && pkey->type == NID_id_GostR3410_2001) {
> > -        len = 64;
> > +    if (n == 64 && (pkey->type == NID_id_GostR3410_2001 ||
> > +                    pkey->type == NID_id_GostR3410_2012_256)) {
> > +         len = 64;
> > +    } else if (n == 128 && (pkey->type == NID_id_GostR3410_2012_512)) {
> > +         len = 128;
>
> Please check this workaround is actually needed for GOST12
>
>
>
I asked my former colleagues to make extra checks.I will report as soon as
I get the result.

-- 
SY, Dmitry Belyavsky

File: ssl/s3_clnt.c [hunk 3]
```diff
@@ -3090,17 +3108,23 @@ int ssl3_send_client_verify(SSL *s)
...
-                                                 NID_id_GostR3411_94, data);
-            if (EVP_PKEY_sign(pctx, signbuf, &sigsize, data, 32) <= 0) {
+            size_t sigsize =
+                (pkey->type == NID_id_GostR3410_2012_512) ? 128 : 64;
+            int dgst_nid = NID_undef;
+
+            EVP_PKEY_get_default_digest_nid(pkey, &dgst_nid);
+            s->method->ssl3_enc->cert_verify_mac(s, dgst_nid, data);
+            if (EVP_PKEY_sign(pctx, signbuf, &sigsize, data,
+                              EVP_MD_size(EVP_get_digestbynid(dgst_nid))) <= 0) {
```
Comment (mattcaswell) [inline ssl/s3_clnt.c]: I think I'm still not getting something here. You set data to be EVP_MAX_MD_SIZE * 2 in  length, but here you are using `EVP_MD_size(EVP_get_digestnid(dgst_nid))` for the amount of data actually used. By definition EVP_MD_size must be less than EVP_MAX_MD_SIZE right? So why do you need double the amount? In your email to me you said you need to calculate 3 digests: md_gost94 (32 bytes), md_gost_2012_256 (32 bytes) and  md_gost_2012_512 (64 bytes). But where is that done? Am I missing something?

Comment (beldmit) [inline ssl/s3_clnt.c]: Well, it is done at server side. In the request it is in ssl/s3_srvr.c, the part between lines 3034-3049

Comment (beldmit) [inline ssl/s3_clnt.c]: Oh, I got it. My fault. EVP_MAX_MD_SIZE will be enough.

File: include/openssl/ssl.h [hunk 4]
```diff
@@ -218,6 +218,7 @@ extern "C" {
 # define SSL_TXT_aPSK            "aPSK"
 # define SSL_TXT_aGOST94 "aGOST94"
 # define SSL_TXT_aGOST01 "aGOST01"
+# define SSL_TXT_aGOST12 "aGOST12"
 # define SSL_TXT_aGOST  "aGOST"
```
Comment (mattcaswell) [inline include/openssl/ssl.h]: Something strange going on with the alignment here (the strings are out of alignment with all the others in this file). Or did github just mess this up?

**Comment by beldmit** [comment]
It seems to me these strings are misaligned in current version.
https://git.openssl.org/gitweb/?p=openssl.git;a=blob_plain;f=include/openssl/ssl.h;hb=HEAD

File: include/openssl/ssl3.h [hunk 5]
```diff
@@ -357,7 +357,7 @@ extern "C" {
  * SSL3_CT_NUMBER is used to size arrays and it must be large enough to
  * contain all of the cert types defined either for SSLv3 and TLSv1.
  */
-# define SSL3_CT_NUMBER                  9
+# define SSL3_CT_NUMBER                  10
```
Comment (mattcaswell) [inline include/openssl/ssl3.h]: I've been trying to figure this number out because it just doesn't make sense. Before your additions there are 7 certificates types defined here plus an additional 4 in tls1.h (excluding repeats) - so that means the number should have been 11 not 9.

So I did a bit more digging. It turns out that the comment above this definition is completely misleading. The value SSL3_CT_NUMBER is used to size the tmp.ctype array in the SSL structure. This is used to record the certificate types that a server is willing to accept when a client receives a CertificateRequest message. Historically that was the only place we stored the acceptable certificate types - so if we received more than would fit in the array then we would have a problem. Therefore it would be more accurate if the comment actually said:
"...it must be large enough to contain all of the cert types the are currently defined or will ever be defined in any current or future RFC".

In other words it has nothing to do with the number of cert types that OpenSSL itself knows about. That is of course completely ridiculous and unworkable, so some time ago Steve implemented a "fix" which basically means that if we receive more certificate types than will fit into the array then we allocate some dynamic memory and store them there instead.

The result of all of that is I think we should leave SSL3_CT_NUMBER and TLS_CT_NUMBER (which isn't used *anywhere*) alone. It looks like there needs to be some separate work to clean it up - but that's not for you to worry about in this patch.

File: include/openssl/ssl.h (hunk 4 above)
Comment (mattcaswell) [inline include/openssl/ssl.h]: Well since its GOST related anyway, could you fix it as part of this patch?

Comment (beldmit) [inline include/openssl/ssl.h]: Sure.

File: include/openssl/ssl3.h (hunk 5 above)
Comment (beldmit) [inline include/openssl/ssl3.h]: So should I remove changing it from my patch?

Comment (mattcaswell) [inline include/openssl/ssl3.h]: Yes

File: ssl/s3_srvr.c (hunk 2 above)
Comment (beldmit) [inline ssl/s3_srvr.c]: Will be fixed in the next version of the patch.

File: ssl/ssl_ciph.c [hunk 6]
```diff
@@ -1979,8 +2036,9 @@ int ssl_cipher_get_cert_index(const SSL_CIPHER *c)
         return SSL_PKEY_DSA_SIGN;
     else if (alg_a & SSL_aRSA)
         return SSL_PKEY_RSA_ENC;
-    else if (alg_a & SSL_aGOST01)
-        return SSL_PKEY_GOST01;
+    else if (alg_a & (SSL_aGOST01 | SSL_aGOST12))
+        return SSL_PKEY_GOST_EC;
+
```
Comment (mattcaswell) [inline ssl/ssl_ciph.c]: Is this right? As I understood things, for GOST12 you can use GOST12_512, GOST12_256 or GOST01 certs...*but* for GOST01 surely you can *only* use a GOST01 cert? So wouldn't that be:
````
else if (alg_a & SSL_aGOST12)
    return SSL_PKEY_GOST_EC;
else if (alg_a & SSL_aGOST01)
    return SSL_PKEY_GOST_01;
````

Comment (beldmit) [inline ssl/ssl_ciph.c]: Yes, your variant is better. I'll apply it.

File: ssl/s3_enc.c [hunk 7]
```diff
@@ -537,7 +537,7 @@ int ssl3_digest_cached_records(SSL *s, int keep)
         }
 
         /* Loop through bits of algorithm2 field and create MD_CTX-es */
-        for (i = 0; ssl_get_handshake_digest(i, &mask, &md); i++) {
+        for (i = 0; ssl_get_certverify_digest(i, &mask, &md); i++) {
```
Comment (mattcaswell) [inline ssl/s3_enc.c]: I don't understand what you are trying to do with this change. Can you explain it?

File: ssl/s3_srvr.c (hunk 2 above)
Comment (beldmit) [inline ssl/s3_srvr.c]: These lines of code could be removed for the 2012 algorithms.

File: ssl/s3_enc.c (hunk 7 above)
Comment (beldmit) [inline ssl/s3_enc.c]: This is a list of digest to be stored in the tmp.cert_verify_md field. In ssl3_get_cert_verify we select the digest we need according to the pkey we got from the peer.

The ssl_get_handshake_digest can't be used for our purposes because it is used in PRF calculation, and we have only one digest for PRF but a set of them for cert_verify.

Comment (mattcaswell) [inline ssl/s3_enc.c]: The PRF MACs and the handshake MACs are specified independently. So for GOST2012-GOST8912-GOST8912, you have defined a PRF MAC of TLS1_PRF_GOST12_256, whilst you have additionally specified a handshake MAC of SSL_HANDSHAKE_MAC_GOST12_256. Similarly for GOST2012-NULL-GOST12.

Comment (beldmit) [inline ssl/s3_enc.c]: The function I have introduced is not for the handshake mac but for the cert_verify mac.

Comment (beldmit) [inline ssl/s3_enc.c]: I want to ensure that any digest necessary for cert_verify is calculated. I was not sure how to implement it so I provided a separate function for enumerating them.

File: ssl/s3_lib.c [hunk 8]
```diff
@@ -3770,6 +3769,34 @@ OPENSSL_GLOBAL const SSL_CIPHER ssl3_ciphers[] = {
...
+     1,
+     "GOST2012-GOST8912-GOST8912",
+     0x0300ff85,
+     SSL_kGOST,
+     SSL_aGOST12 | SSL_aGOST01,
+     SSL_eGOST2814789CNT12,
+     SSL_GOST89MAC12,
+     SSL_TLSV1,
+     SSL_NOT_EXP | SSL_HIGH,
+     SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_STREAM_MAC,
```
Comment (mattcaswell) [inline ssl/s3_lib.c]: I think this line is wrong. I think what you want is:
````
SSL_HANDSHAKE_MAC_GOST | TLS1_PRF_GOST12_256 | TLS1_STREAM_MAC
````
I'm fairly sure that will give you the digests that you need for the cert verify. Similarly for cipher below.

File: ssl/s3_enc.c (hunk 7 above)
Comment (mattcaswell) [inline ssl/s3_enc.c]: This is not the way to do it. See my comment below on the cipher definitions in s3_lib.c

File: ssl/s3_srvr.c [hunk 9]
```diff
@@ -3012,19 +3024,32 @@ int ssl3_get_cert_verify(SSL *s)
...
+         *    - client cert 34.10-2012 512 with 64-bytes digest offset 64
+         */
+        size_t offset = (pkey->type == NID_id_GostR3410_2012_512) ? 64
+            : ((pkey->type == NID_id_GostR3410_2012_256) ? 32 : 0);
+        size_t idx;
         EVP_PKEY_CTX *pctx = EVP_PKEY_CTX_new(pkey, NULL);
         EVP_PKEY_verify_init(pctx);
-        if (len != 64) {
+        if (len != sigsize) {
             fprintf(stderr, "GOST signature length is %d", len);
```
Comment (mattcaswell) [inline ssl/s3_srvr.c]: This needs to log an error and goto f_err with al set to SSL_AD_DECODE_ERROR. There's a potential buffer overrun here otherwise

File: ssl/ssl_ciph.c [hunk 10]
```diff
@@ -258,18 +266,30 @@ static int ssl_cipher_info_find(const ssl_cipher_table * table,
...
+    SSL_HANDSHAKE_MAC_SHA384, SSL_HANDSHAKE_MAC_GOST12_256, 0,
+    SSL_HANDSHAKE_MAC_GOST12_512,
+};
+
+static int ssl_certverify_digest_flag[SSL_MD_NUM_IDX] = {
+    SSL_HANDSHAKE_MAC_MD5, SSL_HANDSHAKE_MAC_SHA,
+    SSL_HANDSHAKE_MAC_GOST, 0, SSL_HANDSHAKE_MAC_SHA256,
+    SSL_HANDSHAKE_MAC_SHA384, SSL_HANDSHAKE_MAC_GOST, 0,
+    SSL_HANDSHAKE_MAC_GOST,
 };
```
Comment (mattcaswell) [inline ssl/ssl_ciph.c]: As per comments above, you shouldn't need this array.

File: ssl/ssl_ciph.c [hunk 11]
```diff
@@ -494,7 +520,11 @@ void ssl_load_ciphers(void)
         if (md == NULL) {
             disabled_mac_mask |= t->mask;
         } else {
-            ssl_mac_secret_size[i] = EVP_MD_size(md);
+            if ((t->nid == NID_id_Gost28147_89_MAC) || (t->nid == NID_gost_mac_12)) {
+                ssl_mac_secret_size[i] = 32;
```
Comment (mattcaswell) [inline ssl/ssl_ciph.c]: The previous code did this on line 567 below. It would be nice to be consistent

File: ssl/ssl_ciph.c [hunk 12]
```diff
@@ -707,6 +741,19 @@ int ssl_get_handshake_digest(int idx, long *mask, const EVP_MD **md)
...
+        return 0;
+    }
+    *mask = ssl_certverify_digest_flag[idx];
+    if (*mask)
+        *md = ssl_digest_methods[idx];
+    else
+        *md = NULL;
+    return 1;
+}
+
```
Comment (mattcaswell) [inline ssl/ssl_ciph.c]: Not needed as per comments above.

File: ssl/ssl_ciph.c [hunk 13]
```diff
@@ -1979,8 +2036,11 @@ int ssl_cipher_get_cert_index(const SSL_CIPHER *c)
         return SSL_PKEY_DSA_SIGN;
     else if (alg_a & SSL_aRSA)
         return SSL_PKEY_RSA_ENC;
+		else if (alg_a & SSL_aGOST12)
```
Comment (mattcaswell) [inline ssl/ssl_ciph.c]: Something strange with formatting here

File: ssl/ssl_locl.h [hunk 14]
```diff
@@ -1874,6 +1892,7 @@ __owur int ssl_cipher_get_evp(const SSL_SESSION *s, const EVP_CIPHER **enc,
                        const EVP_MD **md, int *mac_pkey_type,
                        int *mac_secret_size, SSL_COMP **comp, int use_etm);
 __owur int ssl_get_handshake_digest(int i, long *mask, const EVP_MD **md);
+__owur int ssl_get_certverify_digest(int i, long *mask, const EVP_MD **md);
```
Comment (mattcaswell) [inline ssl/ssl_locl.h]: Not needed

File: ssl/t1_lib.c [hunk 15]
```diff
@@ -954,6 +954,9 @@ static const unsigned char tls12_sigalgs[] = {
         tlsext_sigalg(TLSEXT_hash_sha256)
         tlsext_sigalg(TLSEXT_hash_sha224)
         tlsext_sigalg(TLSEXT_hash_sha1)
+    TLSEXT_hash_gostr3411, TLSEXT_signature_gostr34102001,
+    TLSEXT_hash_gostr34112012_256, TLSEXT_signature_gostr34102012_256,
+    TLSEXT_hash_gostr34112012_512, TLSEXT_signature_gostr34102012_512
```
Comment (mattcaswell) [inline ssl/t1_lib.c]: This will mean we will advertise support for these algorithms even when there is no GOST engine

File: ssl/t1_lib.c [hunk 16]
```diff
@@ -3238,7 +3267,11 @@ static const tls12_hash_info tls12_md_info[] = {
     {NID_sha224, 112, EVP_sha224},
     {NID_sha256, 128, EVP_sha256},
     {NID_sha384, 192, EVP_sha384},
-    {NID_sha512, 256, EVP_sha512}
+    {NID_sha512, 256, EVP_sha512},
+		/* FIXME correct secbits values to be found out*/
+		{NID_id_GostR3411_94, 128, md_gost94},
```
Comment (mattcaswell) [inline ssl/t1_lib.c]: There's still a FIXME here and the formatting is wrong

File: ssl/t1_lib.c (hunk 15 above)
Comment (mattcaswell) [inline ssl/t1_lib.c]: Although I just tested this and it doesn't seem to. Not sure why???

Comment (beldmit) [inline ssl/t1_lib.c]: In fact I do not understand what to write there. It will take some time to organize a research.

File: ssl/s3_lib.c (hunk 8 above)
Comment (beldmit) [inline ssl/s3_lib.c]: After backporting the fix you have suggested to 1.2, I get an error on new ciphersuites:

140033947571856:error:1408C095:SSL routines:ssl3_get_finished:digest check failed:s3_both.c:273:
140033947571856:error:140780E5:SSL routines:ssl23_read:ssl handshake failure:s23_lib.c:137:

I still do not remember exactly, why did I introduce a separate function. I will study it again.

Comment (beldmit) [inline ssl/s3_lib.c]: I got it!

The problem is in the tls1_final_finish_mac function. It calls the ssl_handshake_hash function calling the ssl_get_handshake_digest.

In the variant you suggest all the GOST hashes are calculated and concatenated. In the variant I suggest we calculate only the necessary one.

Comment (mattcaswell) [inline ssl/s3_lib.c]: Ahhh. Ok. I think I understand now. So the acceptable hash algorithms in the cert verify could be any of the GOST hashes. However the handshake hash is always a specifically defined one. Am I correct?

If so, I think a possible answer is to build on the SIGALGS code. Normally we calculate all the digests that we're going to need for the Finished message as we go and discard the handshake messages once we've processed them (so there is no opportunity to go back and calculate a different digest). With SIGALGS though if the SIGALGS extension is present we retain all the handshake messages received so far in `s->s3->handshake_buffer`. Only after the CertVerify do we get rid of that and start calculating the digests as we go. So in outline I think the changes you would have to make would be to:

* Revert the above line back to:
````
SSL_HANDSHAKE_MAC_GOST12_256 | TLS1_PRF_GOST12_256 | TLS1_STREAM_MAC,
````
* In s3_srvr.c in function ssl3_accept in the SSL3_ST_SR_KEY_EXCH_A/B state where there is a test for SIGALGS:
````
            } else if (SSL_USE_SIGALGS(s)) {
                s->state = SSL3_ST_SR_CERT_VRFY_A;
                s->init_num = 0;
                if (!s->session->peer)
                    break;
                if (!s->s3->handshake_buffer) {
                    SSLerr(SSL_F_SSL3_ACCEPT, ERR_R_INTERNAL_ERROR);
                    s->state = SSL_ST_ERR;
                    return -1;
                }
                /*
                 * For sigalgs freeze the handshake buffer. If we support
                 * extms we've done this already so this is a no-op
                 */
                if (!ssl3_digest_cached_records(s, 1)) {
                    s->state = SSL_ST_ERR;
                    return -1;
                }
````
We additionally want to take that branch if we are using a GOST algorithm so add an OR clause to the if statement. So what we are doing here is freezing the state of s->s3->handshake_buffer ready for later use by the cert verify code, and initialising the digests needed to calculate the Finished mac.
* In ssl3_get_cert_verify we need to change GOST code to work in a similar way to the SIGALGS code just a bit above it, i.e. this code:
````
    if (SSL_USE_SIGALGS(s)) {
        long hdatalen = 0;
        void *hdata;
        hdatalen = BIO_get_mem_data(s->s3->handshake_buffer, &hdata);
        if (hdatalen <= 0) {
            SSLerr(SSL_F_SSL3_GET_CERT_VERIFY, ERR_R_INTERNAL_ERROR);
            al = SSL_AD_INTERNAL_ERROR;
            goto f_err;
        }
#ifdef SSL_DEBUG
        fprintf(stderr, "Using TLS 1.2 with client verify alg %s\n",
                EVP_MD_name(md));
#endif
        if (!EVP_VerifyInit_ex(&mctx, md, NULL)
            || !EVP_VerifyUpdate(&mctx, hdata, hdatalen)) {
            SSLerr(SSL_F_SSL3_GET_CERT_VERIFY, ERR_R_EVP_LIB);
            al = SSL_AD_INTERNAL_ERROR;
            goto f_err;
        }

        if (EVP_VerifyFinal(&mctx, data, len, pkey) <= 0) {
            al = SSL_AD_DECRYPT_ERROR;
            SSLerr(SSL_F_SSL3_GET_CERT_VERIFY, SSL_R_BAD_SIGNATURE);
            goto f_err;
        }
    }
````
I hope I've understood correctly and that makes some kind of sense :-)

Comment (mattcaswell) [inline ssl/s3_lib.c]: Oh, I missed a spot you'll need to change in ssl3_get_client_hello:
````
    if (!SSL_USE_SIGALGS(s) || !(s->verify_mode & SSL_VERIFY_PEER)) {
        if (!ssl3_digest_cached_records(s, 0))
            goto f_err;
    }
````
The if statement needs to check "if (we're not using SIGALGS AND we're not GOST) OR (we're not verifying the peer) then..."

File: ssl/t1_lib.c [hunk 17]
```diff
@@ -3238,7 +3267,10 @@ static const tls12_hash_info tls12_md_info[] = {
     {NID_sha224, 112, EVP_sha224},
     {NID_sha256, 128, EVP_sha256},
     {NID_sha384, 192, EVP_sha384},
-    {NID_sha512, 256, EVP_sha512}
+    {NID_sha512, 256, EVP_sha512},
+    {NID_id_GostR3411_94,       128, md_gost94},
+    {NID_id_GostR3411_2012_256, 128, md_gost2012_256},
+    {NID_id_GostR3411_2012_512, 256, md_gost2012_512},
```
Comment (mattcaswell) [inline ssl/t1_lib.c]: This is a problem. The function tls12_get_hash_info matches the values TLSEXT_hash_* to rows in this table. TLSEXT_hash_md5 has a value 1 and corresponds to row 0, TLSEXT_hash_sha1 has a value 2 and corresponds to row 1, etc. This is fine but the GOST hashes skip a load of numbers so TLSEXT_hash_gostr3411 has a value 237, so the algorithm in tls12_get_hash_info no longer works.

File: ssl/t1_lib.c (hunk 15 above)
Comment (mattcaswell) [inline ssl/t1_lib.c]: The reason these don't appear when I expected them to is due to the problem I just commented on with the tls12_get_hash_info function.

File: ssl/t1_lib.c (hunk 17 above)
Comment (beldmit) [inline ssl/t1_lib.c]: I can organize a lookup table here or just make an array , but i still do not understand how to deal with the previous place near the line 957.

File: ssl/t1_lib.c (hunk 15 above)
Comment (beldmit) [inline ssl/t1_lib.c]: Btw the comment at line 923 is 

/*
 * List of supported signature algorithms and hashes. Should make this
 * customisable at some point, for now include everything we support.
 */

And how did you test the list of supported algorithms?

Comment (mattcaswell) [inline ssl/t1_lib.c]: I config'd with enable-ssl-trace, and then used the -trace option with s_client so that I could look at what sigalgs were sent in the relevant ClientHello extension.

File: ssl/t1_lib.c (hunk 17 above)
Comment (mattcaswell) [inline ssl/t1_lib.c]: I think some kind of lookup table is the way to go. As to previous location...tricky. We could have two lists - one with all the standard algs (as now) and one with both the standard algs and the gost ones. Really we need to rewrite that function completely...but that's probably not for you to worry about as part of this patch.

File: ssl/s3_lib.c (hunk 8 above)
Comment (beldmit) [inline ssl/s3_lib.c]: Dear Matt, 
Unfortunately it does not work this way. 
Now I get an error in the tls1_cert_verify_mac function at line 709.

It is possible to add the GOST extra hashes manually, but it will be the other way to do what I am doing introducing the ssl_get_certverify_digest function. I am not sure it's a good idea.

Comment (mattcaswell) [inline ssl/s3_lib.c]: Is your amended code somewhere I can look at? Also how can I recreate the issue you encountered?

Comment (beldmit) [inline ssl/s3_lib.c]: Here is the patch against my fork (https://github.com/beldmit/openssl/tree/tls_gost_2012)
--- a/ssl/s3_enc.c
+++ b/ssl/s3_enc.c
@@ -537,7 +537,8 @@ int ssl3_digest_cached_records(SSL *s, int keep)
         }
 
         /* Loop through bits of algorithm2 field and create MD_CTX-es */
-        for (i = 0; ssl_get_certverify_digest(i, &mask, &md); i++) {
+        /* for (i = 0; ssl_get_certverify_digest(i, &mask, &md); i++) { */ 
+        for (i = 0; ssl_get_handshake_digest(i, &mask, &md); i++) {
             if ((mask & ssl_get_algorithm2(s)) && md) {
                 s->s3->handshake_dgst[i] = EVP_MD_CTX_create();
                 if (EVP_MD_nid(md) == NID_md5) {
diff --git a/ssl/s3_srvr.c b/ssl/s3_srvr.c
index ebbdebd..08c0e2e 100644
--- a/ssl/s3_srvr.c
+++ b/ssl/s3_srvr.c
@@ -579,7 +579,7 @@ int ssl3_accept(SSL *s)
                  */
                 s->state = SSL3_ST_SR_CHANGE_A;
                 s->init_num = 0;
-            } else if (SSL_USE_SIGALGS(s)) {
+            } else if (SSL_USE_SIGALGS(s) || (s->s3->tmp.new_cipher->algorithm_auth & (SSL_aGOST12|SSL_aGOST01) )) {
                 s->state = SSL3_ST_SR_CERT_VRFY_A;
                 s->init_num = 0;
                 if (!s->session->peer)
@@ -1427,7 +1427,7 @@ int ssl3_get_client_hello(SSL *s)
         s->s3->tmp.new_cipher = s->session->cipher;
     }
 
-    if (!SSL_USE_SIGALGS(s) || !(s->verify_mode & SSL_VERIFY_PEER)) {
+    if (!(SSL_USE_SIGALGS(s) || (s->s3->tmp.new_cipher->algorithm_auth & (SSL_aGOST12|SSL_aGOST01) ) )|| !(s->verify_mode & SSL_VERIFY_PEER)) {
         if (!ssl3_digest_cached_records(s, 0))
             goto f_err;
     }
@@ -2832,7 +2832,7 @@ int ssl3_get_cert_verify(SSL *s)
     if (n == 64 && (pkey->type == NID_id_GostR3410_2001)) {
          len = 64;
     } else {
-        if (SSL_USE_SIGALGS(s)) {
+        if (SSL_USE_SIGALGS(s) || (s->s3->tmp.new_cipher->algorithm_auth & (SSL_aGOST12|SSL_aGOST01) )) {
             int rv;
 
             if (!PACKET_get_bytes(&pkt, &sig, 2)) {
@@ -2869,7 +2869,7 @@ int ssl3_get_cert_verify(SSL *s)
         goto f_err;
     }
 
-    if (SSL_USE_SIGALGS(s)) {
+    if (SSL_USE_SIGALGS(s) || (s->s3->tmp.new_cipher->algorithm_auth & (SSL_aGOST12|SSL_aGOST01) )) {
         long hdatalen = 0;
         void *hdata;
         hdatalen = BIO_get_mem_data(s->s3->handshake_buffer, &hdata);


It breaks self-compatibility 
(
s_client -connect localhost:4433 -CAfile sslCA/cacert.pem -verify_return_error -verify 1 -state -cert ssl_user_gost2012_256_B/cert.pem -key ssl_user_gost2012_256_B/seckey.pem -ign_eof
s_server -msg -cert localhost_gost2001_XA/cert.pem -key localhost_gost2001_XA/seckey.pem -Verify 3 -verify_return_error -www -CAfile sslCA/cacert.pem -tls1

Diagnostics:
47070142607912:error:1414D172:SSL routines:tls12_check_peer_sigalg:wrong signature type:t1_lib.c:1017:
47070142607912:error:140940E5:SSL routines:ssl3_read_bytes:ssl handshake failure:record/rec_layer_s3.c:1034:
)

and compatibility with the 3rd-party implementation
(
s_client -msg -cipher GOST2012-GOST8912-GOST8912 -cert U_x_gost2001_B/cert.pem -key U_x_gost2001_B/seckey.pem -CAfile ca_gost2012_512.pem -verify_return_error -verify 1 -connect v-cp4-12L.vm.cryptocom.ru:443 -ign_eof

47777785954856:error:1411E144:SSL routines:tls1_cert_verify_mac:no required digest:t1_enc.c:709:
47777785954856:error:140940E5:SSL routines:ssl3_read_bytes:ssl handshake failure:record/rec_layer_s3.c:1260:
)

File: ssl/t1_lib.c (hunk 15 above)
Comment (beldmit) [inline ssl/t1_lib.c]: Now they are present in the list provided by the client, though they are marked as UNKNOWN. Is there anything left to do here?

Comment (mattcaswell) [inline ssl/t1_lib.c]: Are they also NOT present if the GOST engine is not available? Also see `ssl_sig_tbl` and `ssl_md_tbl` in t1_trce.c. These need to be updated I think to get rid of the "UNKNOWN" in the trace output.

Comment (beldmit) [inline ssl/t1_lib.c]: Fixed now.

File: ssl/s3_lib.c (hunk 8 above)
Comment (mattcaswell) [inline ssl/s3_lib.c]: Can you send me the various certs/keys etc you are using (I assume they are all test ones) in the above command lines? I will try and recreate this locally.

Comment (beldmit) [inline ssl/s3_lib.c]: Fixed now.

File: ssl/s3_lib.c [hunk 18]
```diff
@@ -3449,7 +3449,6 @@ OPENSSL_GLOBAL const SSL_CIPHER ssl3_ciphers[] = {
      256,
      256},
 #endif
-
```
Comment (mattcaswell) [inline ssl/s3_lib.c]: Spurious line break removal here

File: ssl/statem/statem_clnt.c [hunk 19]
```diff
@@ -1562,11 +1563,22 @@ MSG_PROCESS_RETURN tls_process_server_certificate(SSL *s, PACKET *pkt)
...
+                       SSL_R_WRONG_CERTIFICATE_TYPE);
+                goto f_err;
+            }
+        } else {
+            x = NULL;
+            al = SSL_AD_ILLEGAL_PARAMETER;
+            SSLerr(SSL_F_TLS_PROCESS_SERVER_CERTIFICATE,
+                   SSL_R_WRONG_CERTIFICATE_TYPE);
+            goto f_err;
+        }
```
Comment (mattcaswell) [inline ssl/statem/statem_clnt.c]: Can we collapse this into a single "if"? I don't like that we have the same code repeated twice for the inner "if" and the "else".

**Comment by mattcaswell** [comment]
This is looking good. Nearly there. I just made a couple of minor comments in this pass.

File: ssl/statem/statem_clnt.c (hunk 19 above)
Comment (jrchamp) [inline ssl/statem/statem_clnt.c]: If I have the logic correct, this appears to be the combined if statement:
```c
      if (exp_idx >= 0 && i != exp_idx && (exp_idx != SSL_PKEY_GOST_EC ||
            (i != SSL_PKEY_GOST12_512 && i != SSL_PKEY_GOST12_256 && i != SSL_PKEY_GOST01))) {
```

**Comment by mattcaswell** [comment]
+1

**Comment by richsalz** [comment]
+1

**Comment by mattcaswell** [comment]
This has now been merged.

**Comment by mattcaswell** [comment]
Thanks for the contribution Dmitry! :-)

**Comment by richsalz** [comment]
Yes, Dmitry, thanks very much!  This is great work.
